search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,587 @@
+                break;
+            }
+        }
+    }
+
+    g_strfreev (strs);
+
+    return ret;
+}
+
+static GVariant *
+get_search_results (gchar      **terms,
+                    GHashTable  *page_data)
+{
+    GVariantBuilder results;
+    gchar *term = g_strjoinv (" ", terms);
+    GHashTableIter iter;
+    const gchar *page_id;
+    PageData *data;
+
+    g_hash_table_iter_init (&iter, page_data);
+    g_variant_builder_init (&results, G_VARIANT_TYPE ("as"));
+
+    while (g_hash_table_iter_next (&iter, (gpointer *) &page_id, (gpointer *) &data)) {
+        if (match_word (term, data->title_casefold, data->desc_casefold))
+            g_variant_builder_add (&results, "s", page_id);
+    }
+
+    g_free (term);
+
+    return g_variant_new ("(as)", &results);
+}
+
+typedef GVariant * (*ResultGetter) (gchar      **terms,
+                                    GHashTable  *page_data);
+
+typedef struct
+{
+    GDBusMethodInvocation  *invocation;
+    gchar                 **terms;
+    ResultGetter            result_getter;
+} DelayedResultGetter;
+
+
+static DelayedResultGetter *
+delayed_result_getter_new (YelpSearchProviderApp  *app,
+                           GDBusMethodInvocation  *invocation,
+                           gchar                 **terms,
+                           ResultGetter            result_getter)
+{
+    DelayedResultGetter *delayed = g_new0 (DelayedResultGetter, 1);
+
+    delayed->invocation = g_object_ref (invocation);
+    delayed->terms = g_strdupv (terms);
+    delayed->result_getter = result_getter;
+
+    return delayed;
+}
+
+static void
+delayed_result_getter_free (DelayedResultGetter *delayed)
+{
+    g_object_unref (delayed->invocation);
+    g_strfreev (delayed->terms);
+    g_free (delayed);
+}
+
+static gboolean 
+handle_get_initial_result_set (YelpShellSearchProvider2  *skeleton,
+                               GDBusMethodInvocation     *invocation,
+                               gchar                    **terms,
+                               gpointer                   user_data)
+{
+    DelayedResultGetter *delayed;
+    YelpSearchProviderApp *app = YELP_SEARCH_PROVIDER_APP (user_data);
+
+    if (g_strv_length (terms) < 2 && g_utf8_strlen (terms[0], -1) < 3 ) {
+        g_dbus_method_invocation_return_value (invocation, g_variant_new ("(as)", NULL));
+        return TRUE;
+    }
+
+    if (g_hash_table_size (app->page_data_hash_map) > 0) {
+      g_dbus_method_invocation_return_value (invocation,
+                                             get_search_results (terms,
+                                                                 app->page_data_hash_map));
+      return TRUE;
+    }
+
+    delayed = delayed_result_getter_new (app,
+                                         invocation,
+                                         terms,
+                                         get_search_results);
+
+    g_ptr_array_add (app->delayed_result_getters, (gpointer) delayed);
```

Comment:
No need to cast it to `gpointer` - `gpointer` is a bit special that it accepts whatever pointer.